doc/man/man3/HighScoreComponent.3
.TH "HighScoreComponent" 3 "Jeudi 31 Mars 2016" "Version 1" "Arcade" \" -*- nroff -*-
.ad l
.nh
.SH NAME
HighScoreComponent \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <HighScoreComponent\&.hpp>\fP
.PP
Est dérivée de \fBAComponent\fP\&.
.SS "Types publics"

.in +1c
.ti -1c
.RI "enum \fBComponentsPos\fP { \fBGAMENAME\fP, \fBSCORE\fP, \fBPSEUDO\fP, \fBLABEL\fP }"
.br
.in -1c
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBHighScoreComponent\fP (\fBstd::string\fP \fBconst\fP &gameName, size_t score)"
.br
.ti -1c
.RI "virtual \fB~HighScoreComponent\fP ()"
.br
.ti -1c
.RI "\fBvoid\fP \fBUpdatePseudo\fP (\fBconst\fP \fBint\fP key)"
.br
.ti -1c
.RI "\fBUIComponent\fP \fBconst\fP *\fBconst\fP * \fBgetComponentsToDisplay\fP () \fBconst\fP "
.br
.ti -1c
.RI "\fBbool\fP \fBsubmit\fP ()"
.br
.in -1c
.SS "Attributs publics statiques"

.in +1c
.ti -1c
.RI "static \fBconst\fP \fBstd::string\fP \fBpseudoPlaceholder\fP"
.br
.ti -1c
.RI "static \fBconst\fP \fBstd::string\fP \fBlabelHighscores\fP"
.br
.ti -1c
.RI "static \fBconst\fP size_t \fBcomponentNb\fP = 14"
.br
.ti -1c
.RI "static \fBconst\fP \fBVector2\fP< double > \fBhighscoreDim\fP"
.br
.ti -1c
.RI "static \fBconst\fP \fBVector2\fP< double > \fBhighscorePos\fP"
.br
.in -1c
.SH "Description détaillée"
.PP 
Définition à la ligne 12 du fichier HighScoreComponent\&.hpp\&.
.SH "Documentation des énumérations membres"
.PP 
.SS "enum \fBHighScoreComponent::ComponentsPos\fP"

.PP
\fBValeurs énumérées\fP
.in +1c
.TP
\fB\fIGAMENAME \fP\fP
.TP
\fB\fISCORE \fP\fP
.TP
\fB\fIPSEUDO \fP\fP
.TP
\fB\fILABEL \fP\fP
.PP
Définition à la ligne 22 du fichier HighScoreComponent\&.hpp\&.
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "HighScoreComponent::HighScoreComponent (\fBstd::string\fP \fBconst\fP & gameName, size_t score)"

.PP
Définition à la ligne 15 du fichier HighScoreComponent\&.cpp\&.
.SS "HighScoreComponent::~HighScoreComponent ()\fC [virtual]\fP"

.PP
Définition à la ligne 75 du fichier HighScoreComponent\&.cpp\&.
.SH "Documentation des fonctions membres"
.PP 
.SS "\fBUIComponent\fP \fBconst\fP *\fBconst\fP * HighScoreComponent::getComponentsToDisplay () const"

.PP
Définition à la ligne 108 du fichier HighScoreComponent\&.cpp\&.
.SS "\fBbool\fP HighScoreComponent::submit ()"

.PP
Définition à la ligne 113 du fichier HighScoreComponent\&.cpp\&.
.SS "\fBvoid\fP HighScoreComponent::UpdatePseudo (\fBconst\fP \fBint\fP key)"

.PP
Définition à la ligne 83 du fichier HighScoreComponent\&.cpp\&.
.SH "Documentation des données membres"
.PP 
.SS "\fBconst\fP size_t HighScoreComponent::componentNb = 14\fC [static]\fP"

.PP
Définition à la ligne 19 du fichier HighScoreComponent\&.hpp\&.
.SS "\fBconst\fP \fBVector2\fP< double > HighScoreComponent::highscoreDim\fC [static]\fP"

.PP
Définition à la ligne 20 du fichier HighScoreComponent\&.hpp\&.
.SS "\fBconst\fP \fBVector2\fP< double > HighScoreComponent::highscorePos\fC [static]\fP"

.PP
Définition à la ligne 21 du fichier HighScoreComponent\&.hpp\&.
.SS "\fBconst\fP \fBstd::string\fP HighScoreComponent::labelHighscores\fC [static]\fP"

.PP
Définition à la ligne 18 du fichier HighScoreComponent\&.hpp\&.
.SS "\fBconst\fP \fBstd::string\fP HighScoreComponent::pseudoPlaceholder\fC [static]\fP"

.PP
Définition à la ligne 17 du fichier HighScoreComponent\&.hpp\&.

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour Arcade à partir du code source\&.


src/HighScoreComponent.cpp
//
// Created by gaspar_q on 3/15/16.
//

#include "HighScoreComponent.hpp"
#include "ArcadeSystem.hpp"

const std::string   HighScoreComponent::pseudoPlaceholder("Enter your pseudo");
const std::string   HighScoreComponent::labelHighscores("Highscores");
const Vector2<double>  HighScoreComponent::highscoreDim(15, 20);
const Vector2<double>  HighScoreComponent::highscorePos(7, 5);

static std::string  pseudo(HighScoreComponent::pseudoPlaceholder);

HighScoreComponent::HighScoreComponent(std::string const &gameName, int score) :
    AComponent(HighScoreComponent::highscorePos, AComponent::ComponentColor::WHITE),
    gameName(gameName),
    highScores(gameName + ".arcade"),
    score(score)
{
    const std::vector<const t_highScore *>  scores = highScores.getAllHighScore();

    //Nom du jeu
    components[ComponentsPos::GAMENAME] = new UIComponent(Vector2<double>(0, 0), AComponent::ComponentColor::WHITE, Vector2<double>(0, 0), "");
    //Score
    components[HighScoreComponent::ComponentsPos::SCORE] = new UIComponent(Vector2<double>(0, 0), AComponent::ComponentColor::WHITE, Vector2<double>(0, 0), "");
    //Enter you pseudo
    components[ComponentsPos::PSEUDO] = new UIComponent(
        Vector2<double>(ArcadeSystem::winWidth / 2, 3), AComponent::ComponentColor::YELLOW,
        Vector2<double>(0, 0), HighScoreComponent::pseudoPlaceholder);
    //Highscores label
    components[ComponentsPos::LABEL] = new UIComponent(
        Vector2<double>(ArcadeSystem::winWidth / 2, 5), AComponent::ComponentColor::WHITE,
        Vector2<double>(0, 0), HighScoreComponent::labelHighscores);
    // All scores
    for (size_t i = 0, len = scores.size(); i < 10; i++) {
        if (i < len) {
            components[i + 3] = new UIComponent(
                Vector2<double>(ArcadeSystem::winWidth / 2, (i * 2) + 7), AComponent::ComponentColor::WHITE,
                Vector2<double>(0, 0), scores[i]->getName() + std::string(" ") + std::to_string(scores[i]->getScore()));
        } else {
            components[i + 3] = NULL;
        }
    }
    components[13] = NULL;
}

HighScoreComponent::~HighScoreComponent()
{
    for (size_t i = 0; i < HighScoreComponent::componentNb && components[i] != NULL; ++i) {
        delete(components[i]);
    }
}

void HighScoreComponent::UpdatePseudo(const int key)
{
    if (key == ArcadeSystem::Backspace)
    {
        if (!pseudo.empty() && pseudo != HighScoreComponent::pseudoPlaceholder)
            pseudo = pseudo.substr(0, pseudo.length() - 1);
    }
    else if (key >= ' ' && key <= '~' && pseudo.length() < 20)
    {
        if (pseudo == HighScoreComponent::pseudoPlaceholder)
            pseudo.clear();
        pseudo.push_back(static_cast<char>(key));
    }

    if (pseudo.empty())
        pseudo = HighScoreComponent::pseudoPlaceholder;
    delete(components[HighScoreComponent::ComponentsPos::PSEUDO]);
    components[HighScoreComponent::ComponentsPos::PSEUDO] = new UIComponent(
        Vector2<double>(ArcadeSystem::winWidth / 2, 3), AComponent::ComponentColor::WHITE,
        Vector2<double>(0, 0), pseudo);
}

UIComponent const * const *HighScoreComponent::getComponentsToDisplay() const
{
    return components;
}

bool    HighScoreComponent::submit()
{
    if (pseudo.empty() || pseudo == HighScoreComponent::pseudoPlaceholder) {
        return true;
    }
    highScores.setHighScore(new t_highScore(score, pseudo));
    highScores.flush();
    pseudo = HighScoreComponent::pseudoPlaceholder;
    return true;
}
